src/main/java/org/mule/extension/oauth2/internal/service/OAuthContextServiceAdapter.java

```diff
@@ -0,0 +1,135 @@
+ * service.
+ */
+public final class OAuthContextServiceAdapter {
+
+  private static Class<? extends ResourceOwnerOAuthContext> ctxWithStateClass;
+  private static Constructor<? extends ResourceOwnerOAuthContext> ctxWithStateConstructor;
+  private static Constructor<? extends ResourceOwnerOAuthContext> ctxWithStateCopyConstructor;
+  private static Method createRefreshUserOAuthContextLock = null;
+  private static Method getRefreshUserOAuthContextLock = null;
+  private static Method dancerName = null;
+
+  static {
+    try {
+      ctxWithStateClass = (Class<? extends ResourceOwnerOAuthContext>) Class
+          .forName("org.mule.runtime.oauth.api.state.ResourceOwnerOAuthContextWithRefreshState");
+
+      ctxWithStateConstructor = ctxWithStateClass.getConstructor(String.class);
+      ctxWithStateCopyConstructor = ctxWithStateClass.getConstructor(ResourceOwnerOAuthContext.class);
+
+      getRefreshUserOAuthContextLock =
+          ctxWithStateClass.getDeclaredMethod("getRefreshOAuthContextLock", String.class, LockFactory.class);
+
+      createRefreshUserOAuthContextLock =
+          ctxWithStateClass.getDeclaredMethod("createRefreshOAuthContextLock", String.class, LockFactory.class, String.class);
+
+      dancerName = OAuthDancerBuilder.class.getDeclaredMethod("name", String.class);
+    } catch (NoSuchMethodException | ClassNotFoundException e) {
+      // Nothing to do, this is just using an older version of the api
+    } catch (SecurityException e) {
+      throw e;
+    }
+  }
+
+  private OAuthContextServiceAdapter() {
+    // Nothing to do
+  }
+
+  public static ResourceOwnerOAuthContext createResourceOwnerOAuthContext(String resourceOwnerId, String name,
+                                                                          LockFactory lockFactory) {
+    if (ctxWithStateClass != null) {
+      try {
+        return ctxWithStateConstructor.newInstance(resourceOwnerId);
+      } catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+        throw new MuleRuntimeException(e);
+      }
+    } else {
+      return new DefaultResourceOwnerOAuthContext(createLockForResourceOwner(resourceOwnerId, name, lockFactory),
+                                                  resourceOwnerId);
+    }
+  }
+
+  private static Lock createLockForResourceOwner(String resourceOwnerId, String configName, LockFactory lockFactory) {
+    return lockFactory.createLock(configName + "-" + resourceOwnerId);
+  }
+
+  public static ResourceOwnerOAuthContext migrateContextIfNeeded(ResourceOwnerOAuthContext resourceOwnerOAuthContext, String name,
+                                                                 LockFactory lockFactory) {
+    if (ctxWithStateClass != null) {
+      try {
+        return ctxWithStateCopyConstructor.newInstance(resourceOwnerOAuthContext);
+      } catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
```

Comment:
for `InstantiationException` and `InvocationTargetException` , shouldn't you actually propagate the cause?